ocsp/responder/redis/checked_redis_source.go

```diff
@@ -0,0 +1,107 @@
+package redis
+
+import (
+	"context"
+	"errors"
+	"sync"
+
+	"github.com/go-gorp/gorp/v3"
+	"github.com/letsencrypt/boulder/core"
+	"github.com/letsencrypt/boulder/db"
+	blog "github.com/letsencrypt/boulder/log"
+	"github.com/letsencrypt/boulder/ocsp/responder"
+	"github.com/letsencrypt/boulder/sa"
+	"github.com/prometheus/client_golang/prometheus"
+	"golang.org/x/crypto/ocsp"
+)
+
+// dbSelector is a limited subset of the db.WrappedMap interface to allow for
+// easier mocking of mysql operations in tests.
+type dbSelector interface {
+	SelectOne(holder interface{}, query string, args ...interface{}) error
+	WithContext(ctx context.Context) gorp.SqlExecutor
+}
+
+type checkedRedisSource struct {
+	base    *redisSource
+	dbMap   dbSelector
+	counter *prometheus.CounterVec
+	log     blog.Logger
+}
+
+func NewCheckedRedisSource(base *redisSource, dbMap dbSelector, stats prometheus.Registerer, log blog.Logger) *checkedRedisSource {
+	counter := prometheus.NewCounterVec(prometheus.CounterOpts{
+		Name: "ocsp_db_responses",
+		Help: "Count of OCSP requests/responses by action taken by the dbSource",
+	}, []string{"result"})
+	stats.MustRegister(counter)
+
+	return &checkedRedisSource{
+		base:    base,
+		dbMap:   dbMap,
+		counter: counter,
+		log:     log,
+	}
+}
+
+var ocspCodeToStatus = map[int]core.OCSPStatus{
+	ocsp.Good:    core.OCSPStatusGood,
+	ocsp.Revoked: core.OCSPStatusRevoked,
+}
+
+// Response implements the responder.Source interface. It looks up the requested OCSP
+// response in the redis cluster and looks up the corresponding status in the DB. If
+// the status disagrees with what redis says, it signs a fresh response and serves it.
+func (src *checkedRedisSource) Response(ctx context.Context, req *ocsp.Request) (*responder.Response, error) {
+	serialString := core.SerialToString(req.SerialNumber)
+
+	var wg sync.WaitGroup
+	wg.Add(2)
+	var dbStatus core.CertificateStatus
+	var redisResult *responder.Response
+	var redisErr, dbErr error
+	go func() {
+		defer wg.Done()
+		dbStatus, dbErr = sa.SelectCertificateStatus(src.dbMap.WithContext(ctx), serialString)
+	}()
+	go func() {
+		defer wg.Done()
+		redisResult, redisErr = src.base.Response(ctx, req)
+	}()
+	wg.Wait()
+
+	if dbErr != nil {
+		if db.IsNoRows(dbErr) {
+			src.counter.WithLabelValues("not_found").Inc()
+			return nil, responder.ErrNotFound
+		}
+
+		src.counter.WithLabelValues("db_error").Inc()
+		return nil, dbErr
+	}
+
+	if redisErr != nil {
+		return nil, redisErr
+	}
```

Comment:
If redis hits an error, do we want to fully error out, or do we want to go ahead with signing and serving a fresh response?

Is there an easy way for us to tell the difference between retryable errors (e.g. "the query worked, just didn't find that item") and stop-the-world errors (e.g. "redis is down, please don't flood the CA with signing requests")?

Reply:
The first case ("didn't find that item") is handled a layer down, at the `redisSource`. It turned 404s from Redis into a signAndSave.

So if we hit an error here, it's because Redis is erroring (due to unavailability, or overcapacity, or whatever). I think in that case it makes sense to return an error, though we _could_ go ahead and try to sign in such cases if we expect that the CA's signing capacity is always enough to serve our origin traffic.

Reply:
> we could go ahead and try to sign in such cases if we expect that the CA's signing capacity is always enough to serve our origin traffic.

Do we have a good handle on whether this is the case or not? Would it make sense to limit live signing to ensure that it doesn't over-run our capacity?

Reply:
The liveSource already has that throttling: it's maxInflightSignings.

I believe we currently can handle all of our origin traffic with live signing, but I still think I'd rather pass through Redis errors. It's a surprising failure case if our cache goes down and that triggers a massive increase in signing traffic to the CA.